### ISSUE TITLE
Hypothesis: It's possible to use graph database as the main data storage

### ISSUE THREAD

**Issue by thegostev** [ckan/ckan#7489]
**Problem statement**

SQL database that's used right now for CKAN puts a limitation on amount of CKAN applications. As an example triplestore approach is used widely in Switzerland, thus CKAN with SQL database should be heavily customized (if it's possible) to fit in.

**Hypothesis**

It's possible to use top-3 most popular graph databases for CKAN as main data storage.

**Discovery work**
_Log here everything you've found during the discovery_

**Validation**
_Why the solution is trustworthy? What makes it strong?_

**Questions to consider:**

Is this change going to break current installations?

Can we provide a backwards compatibility?

How easy is gonna be for current implementations to migrate to this new release?

Do current versions of CKAN have the adequate resources/support to migrate to this new version?

Are we going to change the database schema?

Are we going to change the API?

Are we going to deprecate Interfaces?

**Comment by anuveyatsu** [comment]
I would start here with some quick recap about why this is needed at all:

CKAN allows users to have default metadata schema to get started quickly and create necessary information about datasets. Users can also customize the metadata schema and recommended approach is to use https://github.com/ckan/ckanext-scheming extension. In the same time there is a number of metadata standards defined by different organizations:

- DCAT - https://www.w3.org/TR/vocab-dcat-3/
- Dublin Core - https://www.dublincore.org/specifications/dublin-core/
- Frictionless Data - https://frictionlessdata.io/introduction/
- etc.

DCAT has been adopted by number of governments and other organizations including:

- DCAT-AP (EU Gov) - https://ec.europa.eu/isa2/dcat-ap-key-european-data-publishers-reach-data-consumers-and-create-european-open-data-ecosystem_en/ - it also has extensions for geo and stats datasets.
- DCAT-US (US Gov) - https://resources.data.gov/standards/catalog/dcat-us/

Some governments are also creating their own extension of the DCAT:

- Switzerland - https://www.dcat-ap.ch/
- might be others

**Note that CKAN has a popular extension to support DCAT - https://github.com/ckan/ckanext-dcat.** However, it might not be sufficient to support all features of DCAT and it was suggested that CKAN could support graph databases.

**Before we proceed with how we could implement graph database support, we should understand why it is needed and what it is that current CKAN cannot handle in particular.**

**Comment by thegostev** [comment]
Thanks for the update @anuveyatsu, I like how you explained the matter around standards. My impression was that in EU happens data infrastructure change, if it's so and we can gather evidence we can think through how to join this new infrastructure with interoperable set of technologies on the CKAN side. 

I've restructured the template so that we're separately discovering the need and solutions to this need.

**Comment by thegostev** [comment]
He's the insight by Cheng-Jen, published with his consent:

We have carried out some experiments on linked data (http://data.odw.tw/; source code: https://gitlab.com/iislod) with CKAN several years ago. However, we did not actually connect CKAN with any graph databases. Instead, we [converted](https://gitlab.com/iislod/dat2ld) the original data to triples first, then imported (harvested) those triples into CKAN and [OpenLink Virtuoso](https://virtuoso.openlinksw.com/) (a famous triple store) respectively to acquire [a SPARQL endpoint](http://data.odw.tw/sparql). Please find the slides presented at the [CKANCon 2016](https://web.archive.org/web/20161014102210/http://ckan.org/ckancon-2016/) below:

https://m.odw.tw/u/odw/m/20161004-ckancon-sls/

Therefore, it would be great if we can have a separate triple store with a SPARQL endpoint in CKAN 3.0. And I would be happy to help.

Best,

Cheng-Jen

**Comment by u10313335** [comment]
> He's the insight by Cheng-Jen from 🇹🇼, published with his consent:
> 
> We have carried out some experiments on linked data (http://data.odw.tw/; source code: https://gitlab.com/iislod) with CKAN several years ago. However, we did not actually connect CKAN with any graph databases. Instead, we [converted](https://gitlab.com/iislod/dat2ld) the original data to triples first, then imported (harvested) those triples into CKAN and [OpenLink Virtuoso](https://virtuoso.openlinksw.com/) (a famous triple store) respectively to acquire [a SPARQL endpoint](http://data.odw.tw/sparql). Please find the slides presented at the [CKANCon 2016](https://web.archive.org/web/20161014102210/http://ckan.org/ckancon-2016/) below:
> 
> https://m.odw.tw/u/odw/m/20161004-ckancon-sls/
> 
> Therefore, it would be great if we can have a separate triple store with a SPARQL endpoint in CKAN 3.0. And I would be happy to help.

Hello everyone,

I'm the one who provided the above comments.
And thank you @thegostev for identifying the issue about graph database.

**Comment by amercader** [comment]
My thoughts on this are mostly influenced by having worked on ckanext-dcat to support different DCAT-AP profiles.

Just to get this out of the way I personally think that there is no sense in replacing the main PostgresSQL database with a graph database as was originally suggested in this issue. It works great for all CKAN needs, it's battle tested and well supported, and much easier to manage than a triple store.

Having said that, if the goal is be able to have a semantic-ready triple store (of the data or the metadata) that can have arbitrary fields linked to a particular datasets, and that is synchronized with the CKAN model, I think a decoupled approach based on an extension that hooks at the relevant interfaces could work well.

* For metadata, there are two approaches I can see:

  1. Map CKAN model fields to DCAT triples and viceversa, store dcat:Datasets as normal CKAN datasets and resources and expose them as RDF serializations. This is what ckanext-dcat does. There is a [mapping between CKAN and DCAT entities and fields](https://github.com/ckan/ckanext-dcat#rdf-dcat-to-ckan-dataset-mapping) that users can expand via custom profiles. Other than CKAN default dataset fields these are mostly stored as extras. This is lossy in the sense that of course any field not defined in the mapping would not be persisted (as in every transformation between metadata formats) but it means that all defined fields are stored as part of the CKAN metadata, allowing searches, API access etc
  2. Keep a completely separate triple store, synchronized at the relevant points. That could be when creating/editing/deleting a dataset with `IPackageController`, or when harvesting remote DCAT-AP serializations with a custom harvester that both creates a CKAN dataset (using ckanext-dcat's [DCAT->CKAN parser](https://github.com/ckan/ckanext-dcat/blob/618928be5a211babafc45103a72b6aab4642e964/ckanext/dcat/profiles.py#L915)) and pushes all the triples relevant to the dataset to Virtuoso or a similar database. Of course we would need to keep consistent URIs and identifiers to match both systems but essentially they would work separately. The extension could expose the SPARQL endpoint as part of the CKAN site.

* For data, as @EricSoroos suggested we just need an extension with the exact same hooks as datapusher and xloader that transforms the data to a semantic format that the triple store likes and pushes it when a resource is created or updated, only to a triple store rather than the standard DataStore. Or maybe even a custom `IDataStoreBackend` to power visualizations, data API, etc

Of course these are gross simplifications and there are issues that I have overseen, specially on the semantic side of things where I'm not an expert but overall I think these approaches could work.

But bottom line, I don't think we should approach this as changing the CKAN internals but rather make sure that side-by-side integrations can work fine (which is the general CKAN design approach we are working towards)

**Comment by thegostev** [comment]
Hey @amercader, @EricSoroos, quick status sync. The initial problem discussion led to 2 potential solutions: https://github.com/ckan/ckan/issues/7514 and https://github.com/ckan/ckan/issues/7515. Tech research on these solutions is done, resulting docs are https://hackmd.io/KekGJwn9To6_2AeDs8IStw and https://hackmd.io/Sv8NWNDzTw6bL-n4ZrBT6A?view. 

We'd spent a good portion of time on this problem so we're back to gathering more evidence and finding out more details on exact issues that users have. On the general level now we have an understanding of what can be done, by getting through this thread. So next step is the understanding of priority for this work and finding out more details.

**Comment by EricSoroos** [comment]
TLDR: I think that problems posed here are solvable at the extension level (after proper definition) and I'm -1 on adding a requirement for a graph database to CKAN.

From reading those two posts, there are two somewhat conflicting conclusions:

1) Ckan is able to represent DCAT metadata with an appropriate schema.
2) Ckan doesn't provide representation for `CatalogRecord` and `DataService` classes, and no relationships between `Catalog` and `CatalogRecords`

Also, it appears in both solutions that the core data flow is: `DCAT Harvest -> Repository ->DCAT Republish`, rather than being the source of record for a dataset. For my use cases, we have to do both -- we harvest from various sources, some of which are DCAT, some from other standards, and we have sites where we're the source of record for uploaded or datasets harvested from internal organization sources.  So even if we were to improve our DCAT fidelity to the original by storing an RDF or json-ld blob, we'd still need to be able to generate it from CKAN metadata for those times when we're the originator. 

In either case, we need to be able to publish a DCAT endpoint for EU purposes, and we do that with ckanext-dcat (and ckanext-scheming).

**Comment by bellisk** [comment]
Here are my thoughts, from the perspective of having worked with https://dcat-ap.ch and https://opendata.swiss. We harvest metadata from many data publishers in Switzerland to the portal, which is then harvested in turn by data.europa.eu.

I also don't support requiring a triple store for CKAN to work, but I do think that adding a way to connect one and sync datasets to it (as in @amercader's point `ii`) is a great idea. This would allow us to make RDF metadata from data publishers losslessly available to portals that harvest from us, which is very important to us at the moment. We have also been discussing the possibility of running a triple store alongside CKAN's PostgreSQL database, and it would be more convenient and sustainable to have this as part of an official extension.

So far as I can see, it _should_ be possible to represent the structure of DCAT-AP v2, with the new `CatalogRecord` and `DataService` classes, in CKAN + PostgreSQL, which would let us keep using the `DCAT Harvest -> Repository -> DCAT Republish` workflow that `ckanext-dcat` has. This has the advantages that Adrià described above (searches, API access, etc.) but presents some issues also:

- data is lost if some data publishers start to use an updated DCAT-AP version before its new classes and properties are implemented in `ckanext-dcat`. I feel that this is going to happen soon in Switzerland, and it will be a recurring issue as DCAT-AP and related application profiles are continually updated.
- uncertainty (perhaps just on my part!) as to how these new classes should best be implemented. Should we model `DataService` and `CatalogRecord` as new dataset types? What changes need to be made to the PostgreSQL schema in an extension, in order to actually represent the relationships between all the new and old classes?

I see the value in @amercader's point `i`, but I think making it easy for CKAN sysadmins to run a triple store alongside PostgreSQL is a necessary move so that we can be up-to-date even changes are made to the graph structure that take time to put into place.

**Comment by EricSoroos** [comment]
Are the `DataService` and `CatalogRecord` types fundamentally or topologically different than a dataset? In that, our datasets have metadata and resources, What entities do these have that we can or can't represent?

From my POV, if you're pulling in metadata from a new version or variant schema that you don't have explicit support for in your metadata schema (of whatever form), then you're essentially allowing arbitrary passthrough of that metadata without validation. At that point, it's essentially an opaque blob, either a set of triples, and rdf description, or json-ld. You can't verify it, and you shouldn't modify it because you don't know what it is (at a programming level). If that's the case -- what does a graphdb/triple store gain you over a static blob of RDF/Json-LD?

**Comment by amercader** [comment]
Thanks @bellisk that's really valuable feedback

> Should we model DataService and CatalogRecord as new dataset types?
> Are the DataService and CatalogRecord types fundamentally or topologically different than a dataset?

Disclaimer: It's been a while since I looked at the DCAT spec
Regarding `CatalogRecord`, when starting ckanext-dcat I chose not to model it at all as a new entity in CKAN because in the context of CKAN it was virtually identical to our Datasets (where fields like `metadata_modified` or `metadata_created` would serve the same function as `CatalogRecord.modified` and `CatalogRecord.issued`). I still don't think it would add much value to create separate models for them.

`DataService` is much trickier. According to DCAT `dcat:DataService` exists at the same level as `dcat:Dataset` (they are both sub-classes of a generic `dcat:Resource`, not to confuse with CKAN resources!). In CKAN terms, it could be thought as a CKAN dataset with just one resource for the endpoint URL, but what complicates things is the relationships that DCAT allows between them and `dcat:Distribution` 

```
           ┌──────────────┐                          ┌───────────────────┐
           │              │      dcat:servesDataset  │                   │
           │ dcat:Dataset │◄─────────────────────────┤ dcat:DataService  │
           │              │                          │                   │
           └──────┬───────┘                          └───────────────────┘
                  │                                            ▲
                  │                                            │
                  │                                            │
dcat:distribution │                                            │ dcat:accessService
                  │            ┌───────────────────┐           │
                  │            │                   │           │
                  └───────────►│ dcat:Distribution ├───────────┘
                               │                   │
                               └───────────────────┘
```

In CKAN we would probably represent the same metadata and entities using just one single Dataset, that has different Resources, ones for some files, and others for an API, WMS/INSPIRE endpoint that gives access to the same data. but of course that doesn't align perfectly with the DCAT model, eg we might lose dataset-level fields in DataService.

So perhaps there are different options:
* Decide on some convention like the one above and try to fit DataServices into the standard Dataset/Resources model
* Create custom dataset types for DataServices, and somehow create a strong link between them and their corresponding "Actual Dataset"
* Ignore them entirely, and make clear that anything that is not a dataset or a distribution will only be available on the parallel triple store / blob of RDF

I really don't have a clear idea on this and just trying to move the discussion forward so any other suggestion is welcome

> From my POV, if you're pulling in metadata from a new version or variant schema that you don't have explicit support for in your metadata schema (of whatever form), then you're essentially allowing arbitrary passthrough of that metadata without validation

Just to be clear, when I suggested points `i` and `ii` I was proposing to run both in parallel:

* A CKAN instance imports RDF metadata from one or more sources
* From that input graph, we extract whatever ckanext-dcat supports in a structured way that can get translated to CKAN Dataset and resources (what ckanext-dcat does now). That gives you search, API access, etc
* Besides that, we also import the whole input graph in a way that preserves all the original information
 
Whether we use a triple store or a static dump for the last point I don't know, I honestly don't have enough knowledge to assess how hard it is to maintain, regularly update, etc

> If that's the case -- what does a graphdb/triple store gain you over a static blob of RDF/Json-LD?

I'm not an expert here but I'd imagine that it would allow to provide a SPARQL endpoint thus allowing users familiar with it to perform queries, export metadata, etc

**Comment by amercader** [comment]
@seitenbau-govdata Would be great to have your thoughts on this discussion (from the start! :slightly_smiling_face:  )

**Comment by reicharm** [comment]
Hi,

from our internal discussions at data.gv.at (Austrian government data catalogue) and the discussion with other countries, that also have federated systems, where lots of "smaller" instances are harvested and data is then collectively transferred to data.europa.eu, it will be crucial, to maintain the original rdf metadata. 

We will be in the situation to harvest data in various DCAT-AP formats (GeoDCAT, MobilityDCAT, DCAT-AP.at, ...) and from graph databases as well as data delivered from classic databases or in "classic" json/xml formats and deliver all the aggegrated data to the european data portal, without loosing or altering relevant data in the process of mapping to and from a traditional SQL database. In our developer communities we also see the need for SPARQL endpoints.

I think for us there will be no way around a triple store / graph database and if CKAN will not provide this in the future, this may be a deal breaker with CKAN for us. 

I would also much appreciate it, ifwe could get some kind of rough timeline for the development of CKAN 3.0 because we have to make decisions regarding our future catalogue software in the next few months.

**Comment by Markus92** [comment]
We are having very similar discussions for the Dutch National healthdata catalog, where we'd harvest from multiple regional nodes. Also with an eye on the European Health Data Space, interoperability with other catalogs is major for us. A (semi-)native triple store would be extremely helpful for us.

@reicharm could we reach out to your team for a quick chat on how your team aims to cover these issues? We might have some solutions for each other.

**Comment by thegostev** [comment]
Hi, @reicharm, @Markus92 thanks a lot for contributing to this thread. It's a strong signal for me to prioritize triple store / graph databases support as if you both have the same needs you're not alone and in EU at least it's a need. 

Sorry for idle in responding. In short, to deliver CKAN 3.0 we have to complete a sub-project named UI Revamp to update out-of-the-box UI. At the current stage we're close to form a delivery team backed up by co stewards (Link Digital, Datopian). It'll take approximately 4 months to deliver, so the earliest time is the end of May 2024.

It would be great to have something done by that time on triple store / graph databases support to cover your use cases. As you read the discussion above, the topic isn't easy from the tech standpoint ☎️ I have to understand your use cases in detail to be able to reopen the discussion on solution that'll fulfill them. Let's get in touch. If you could get in contact via alex.gostev@ckan.org I can organize a conversation. Thanks again for voicing over your needs 👍

**Comment by dportoles** [comment]
Thank you for all the contributions on this topic and for generating such valuable knowledge.
Spain has recently launched a new platform for its national open data catalog (datos.gob.es) that supports a national DCAT-AP profile (DCAT-AP-ES, still under administrative review).

The technical approach behind the solution is that publishers submit a catalog in RDF, which is stored in a SPARQL endpoint as a result of the harvesting process. Each harvesting source has an associated profile.
Afterward, SHACL validations are performed according to the previously defined profile.

If the SHACL validations pass, the metadata is interpreted and translated into an internal CKAN data model (based on DCAT-AP, ongoing incremental development). Unknown RDF properties are ignored but stored in the triplestore, so they don’t “get lost.”
This data model is used for filtering datasets, portal functionalities, and other features.

When RDF is generated for the next level (e.g., EDP, ...), the original RDF is reused so that all RDF properties are included, ensuring no loss of information ("lossless" process).

Supporting new DCAT-AP extensions (Geo, Stat, etc.) or new profiles (e.g., Dublin Core) is not a barrier. Scheming properties are always included, and unknown properties are forwarded to future harvesters.
Full support for new profiles requires minimal changes (mainly SHACL validations) and, optionally, schema modifications if needed.

We would be pleased to share our experiences with other initiatives in a similar way!
